### PR TITLE
Video replacement scan for mp4 files

### PR DESCRIPTION
**Pull request by changhiskhan**
1. replacement scan to support queries like "SELECT * FROM 'video.mp4'
2. allow predict to take the video scan output (frame struct) as input for the model (no copying)

I'm wondering if it makes sense to break some of these part as separate extensions, though at least for now it would make release/deployment rather difficult

### CODE REVIEW

**Review by eddyxu — COMMENTED** [review]
Some General questions.

File: integration/duckdb/CMakeLists.txt [hunk 1]
```diff
@@ -37,6 +37,8 @@ if(LANCE_BUILD_PYTORCH)
   endif()
 
   set(BUILD_LIST
+      highgui
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/CMakeLists.txt:40]: actually i don't think highgui is required, only videoio is. We're not statically linking these so they're not that big. The downside is that these are then not portable. In python you probably need to import cv2 or something (similar to torch) but i haven't verified.

File: integration/duckdb/src/lance/duckdb/video_functions.cc [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+// compute the total number of frames
+std::unique_ptr<::duckdb::FunctionData> VideoFunctionBind(
+    ::duckdb::ClientContext& context,
+    ::duckdb::TableFunctionBindInput& input,
+    std::vector<::duckdb::LogicalType>& return_types,
+    std::vector<std::string>& names) {
+  auto bind_data = std::make_unique<VideoFunctionData>();
+  bind_data->uri = input.inputs[0].GetValue<std::string>();
+  // TOOD make it so we only need to open it once per thread
+  auto cap = ::cv::VideoCapture(bind_data->uri);
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/video_functions.cc:94]: VideoCapture can take an https url (and can handle streaming). So at a minimum can do pre-signed s3 urls i believe.

File: integration/duckdb/src/lance/duckdb/video_functions.cc [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+/**
+ * If the table_name ends in .mp4 then implicitly run video_scan
+ */
+::std::unique_ptr<::duckdb::TableFunctionRef> VideoScanReplacement(
+    ::duckdb::ClientContext &context,
+    const ::std::string &table_name,
+    ::duckdb::ReplacementScanData *data) {
+  auto lower_name = ::duckdb::StringUtil::Lower(table_name);
+  if (!::duckdb::StringUtil::EndsWith(lower_name, ".mp4") && !::duckdb::StringUtil::Contains(lower_name, ".mp4?")) {
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/video_functions.cc:37]: oh this is the replacement scan. Basically if the table name ends in .mp4, this tells duckdb to use the 'video_scan' table function.

File: integration/duckdb/src/lance/duckdb/video_functions.cc [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    return nullptr;
+  }
+  auto table_function = ::duckdb::make_unique<::duckdb::TableFunctionRef>();
+  ::std::vector<::std::unique_ptr<::duckdb::ParsedExpression>> children;
+  children.push_back(::duckdb::make_unique<::duckdb::ConstantExpression>(::duckdb::Value(table_name)));
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/video_functions.cc:42]: oh that's a typo should be ::std::make_unique

File: integration/duckdb/src/lance/duckdb/ml/pytorch.cc [hunk 5]
```diff
@@ -152,10 +181,23 @@ std::vector<std::unique_ptr<::duckdb::CreateFunctionInfo>> GetPyTorchFunctions()
...
                                ::duckdb::LogicalType::LIST(::duckdb::LogicalType::FLOAT),
                                Predict));
+  // From frames
+  auto frame = {
+      ::std::pair<::std::string, ::duckdb::LogicalType>("data", ::duckdb::LogicalType::BLOB),
+      ::std::pair<::std::string, ::duckdb::LogicalType>("width", ::duckdb::LogicalType::INTEGER),
+      ::std::pair<::std::string, ::duckdb::LogicalType>("height", ::duckdb::LogicalType::INTEGER),
+      ::std::pair<::std::string, ::duckdb::LogicalType>("channels", ::duckdb::LogicalType::INTEGER)
+  };
+  predict.AddFunction(
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/ml/pytorch.cc:196]: wdym?

File: integration/duckdb/src/lance/duckdb/ml/pytorch.cc [hunk 6]
```diff
@@ -92,6 +92,28 @@ std::optional<cv::Mat> ReadImageFromDuckDBValue(::duckdb::Value image_bytes) {
...
 
+std::optional<cv::Mat> ReadImageFromFrame(::duckdb::Value frame_data,
+                                          ::duckdb::Value frame_rows,
+                                          ::duckdb::Value frame_cols,
+                                          ::duckdb::Value frame_channels) {
+  auto rows = ::duckdb::IntegerValue::Get(frame_rows);
+  auto cols = ::duckdb::IntegerValue::Get(frame_cols);
+  auto channels = ::duckdb::IntegerValue::Get(frame_channels);
+  auto mat_str = ::duckdb::StringValue::Get(frame_data);
+  auto data = (unsigned char*)mat_str.data();
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/ml/pytorch.cc:103]: yeah, i should be able to check the size.

File: integration/duckdb/src/lance/duckdb/ml/pytorch.cc [hunk 7]
```diff
@@ -92,6 +92,28 @@ std::optional<cv::Mat> ReadImageFromDuckDBValue(::duckdb::Value image_bytes) {
...
+
+  // TODO check number of channels
+  cv::Mat img = cv::Mat(rows, cols, CV_8UC3, data);
+  if (img.data == nullptr) {
+    std::cerr << "Failed to parse image: image size=" << mat_str.size() << std::endl;
+    return std::nullopt;
+  }
+  // Convert to float matrix.
+  cv::Mat fmat;
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/ml/pytorch.cc:112]: Depends on what you mean by "normal video processing". At Tubi, for example, video processing was not done using opencv (ffmpeg instead).

File: integration/duckdb/src/lance/duckdb/ml/pytorch.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/ml/pytorch.cc:104]: done

File: integration/duckdb/src/lance/duckdb/video_functions.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/video_functions.cc]: added function docstring

File: integration/duckdb/src/lance/duckdb/video_functions.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+std::vector<std::unique_ptr<::duckdb::CreateTableFunctionInfo>> GetVideoTableFunctions();
+
+/**
+ * Replacement scan for mp4 files (e.g, "SELECT * FROM 'my_video.mp4'")
+ * @param context
+ * @param table_name  this is the video uri
+ * @param data
+ * @return
+ */
+std::unique_ptr<::duckdb::TableFunctionRef> VideoScanReplacement(
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply integration/duckdb/src/lance/duckdb/video_functions.h:39]: I'm a fancy fancy guy

**Pull request closed without merging**